### PR TITLE
feat: fix LinkProps & WithRouter types

### PR DESCRIPTION
**Pull request by felixmosh**
This PR exposes LinkProps to be consumed via `next/link`.

Allow define the type of the query using generics.

`WithRouterProps<{page: string, action: 'delete' | 'edit'}>` makes `query` on router to have an extended type.

### CODE REVIEW

File: packages/next-server/lib/router/router.ts [hunk 1]
```diff
@@ -21,10 +21,10 @@ function toRoute(path: string): string {
   return path.replace(/\/$/, '') || '/'
 }
 
-export type BaseRouter = {
+export type BaseRouter<Query = any> = {
```
Comment (tills13) [inline packages/next-server/lib/router/router.ts]: should this be `Query = {}`. Anything intersected with `any` is `any`.

Comment (felixmosh) [reply packages/next-server/lib/router/router.ts]: Correct, missed it, will change it to `unknown`
image

File: packages/next/client/link.tsx [hunk 2]
```diff
@@ -45,7 +45,7 @@ function formatUrl(url: Url) {
   return url && typeof url === 'object' ? formatWithValidation(url) : url
 }
 
-type LinkProps = {
+export type LinkProps = {
   href: Url
   as?: Url | undefined
```
Comment (Janpot) [inline packages/next/client/link.tsx]: Potentially stupid question, but isn't the ` | undefined` redundant here?

Comment (tills13) [reply packages/next/client/link.tsx]: It is.

**Comment by lfades** [comment]
Hi @felixmosh, Why do you want to type `query`? it won't make your code safer because it relays in the browser URL.

File: packages/next/client/link.tsx (hunk 2 above)
Comment (Timer) [reply packages/next/client/link.tsx]: It doesn't hurt to have for explicitness.

File: packages/next/link.d.ts [hunk 3]
```diff
@@ -1,2 +1,2 @@
 export * from './dist/client/link'
-export { default } from './dist/client/link'
+export { default, LinkProps } from './dist/client/link'
```
Comment (lfades) [inline packages/next/link.d.ts]: This is already exported in the line above and shouldn't be required.

Comment (felixmosh) [reply packages/next/link.d.ts]: Correct, thanx

**Comment by felixmosh** [comment]
> Hi @felixmosh, Why do you want to type `query`? it won't make your code safer because it relays in the browser URL.

Dev experience, in my app, there is page that syncs the some variables with the url, for example page parameter of pagination.

File: packages/next/client/link.tsx (hunk 2 above)
Comment (Janpot) [reply packages/next/client/link.tsx]: It's true that it doesn't hurt, but it's inconsistent with the lines below.
Clean and consistent code also gives me the peace of mind that the authors hold high standards over their code.

Comment (felixmosh) [reply packages/next/client/link.tsx]: Why is that related to my PR? ;)
Do you want me to remove this `undefined`?

Comment (Janpot) [reply packages/next/client/link.tsx]: It's not related to your PR, it just caught my eye while I was reading through the changes.

Comment (felixmosh) [reply packages/next/client/link.tsx]: Cool, no problem to remove it, just a matter of decision :]

**Comment by felixmosh** [comment]
Any updates?

**Pull request closed without merging**